subsys/net/buf.c

```diff
@@ -486,7 +486,6 @@ void net_buf_unref(struct net_buf *buf)
 
 		if (buf->__buf) {
 			data_unref(buf, buf->__buf);
-			buf->__buf = NULL;
 		}
```

Comment:
The problem here is that you're increasing the window where `buf->__buf` has been potentially freed/invalidated by `data_unref()` and the variable actually being set to NULL. I think it might be better if you just move this entire if-block into `net_buf_destroy()`.

Reply:
Moving the if-block into `net_buf_destroy()` essentially required to change `data_unref()`. I just removed `data_unref()` recreating it functionally inside `net_buf_destroy()`.